### PR TITLE
crimson/seastore: add journaling checksumming and nbd based simple perf tester

### PR DESCRIPTION
**Pull request by athanatos**

### CODE REVIEW

**Comment by athanatos** [comment]
9f71b2f converts the uuid encoders to use denc -- that should be ok, right?

File: src/crimson/common/errorator.h [hunk 1]
```diff
@@ -596,14 +596,14 @@ struct errorator {
     }
 
     template <class FuncT>
-    auto finally(FuncT &&func) {
+    auto finally(FuncT &&func) -> std::decay_t<decltype(*this)> {
```
Comment (tchaikov) [inline src/crimson/common/errorator.h:599]: @athanatos i pulled this commit and revised it in #38305. two changes

- use `_future` as the returned type for better readability
- indent the `discard_result()` line with spaces for better readability in GitHub web interface and for better consistency with the rest of this source file.

Comment (tchaikov) [reply src/crimson/common/errorator.h:599]: because this change breaks the build without #38305, i want to get the latter merged first. hence included it in my PR so we can verify that my change does address the FTBFS along with this `finally` fix.

Comment (athanatos) [reply src/crimson/common/errorator.h]: rebased and repushed

File: src/crimson/os/seastore/segment_manager/block.cc [hunk 2]
```diff
@@ -0,0 +1,416 @@
...
+}
+
+BlockSegmentManager::~BlockSegmentManager()
+{
+}
+
+BlockSegmentManager::mount_ret BlockSegmentManager::mount(mount_config_t config)
+{
+  return open_device(
+    config.path, seastar::open_flags::rw
```
Comment (tchaikov) [inline src/crimson/os/seastore/segment_manager/block.cc:277]: does it make sense to pass `open_flags::dsync` as well? so we don't need to bother with the `flush()` call after writing to the storage device.

File: src/crimson/os/seastore/segment_manager/block.cc [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+    return tracker->read_in(
+      device,
+      superblock.tracker_offset
+    ).safe_then([this] {
+      for (segment_id_t i = 0; i < tracker->get_capacity(); ++i) {
+	if (tracker->get(i) == segment_state_t::OPEN) {
+	  tracker->set(i, segment_state_t::CLOSED);
+	}
+      }
+      return tracker->write_out(device, superblock.tracker_offset);
```
Comment (tchaikov) [inline src/crimson/os/seastore/segment_manager/block.cc:296]: i am confused. does this imply that we are able to rewrite a block in-place?

**Review by tchaikov — COMMENTED** [review]
i am still reviewing this changeset.

File: src/crimson/os/seastore/segment_manager/block.cc (hunk 3 above)
Comment (athanatos) [reply src/crimson/os/seastore/segment_manager/block.cc:296]: Yes, BlockSegmentManager uses a conventional device to implement the SegmentManager interface.  This portion simply maintains a control block on the disk recording the state of each segment.  I expect that zns devices will get an analogous ZNSSegmentManager implementation that use the zns library to directly query these states.

File: src/crimson/os/seastore/segment_manager/block.cc (hunk 2 above)
Comment (athanatos) [reply src/crimson/os/seastore/segment_manager/block.cc:277]: I'm not actually sure dsync ensures that a write barrier has been issued to the disk.

File: src/crimson/os/seastore/segment_manager/block.cc [hunk 4]
```diff
@@ -0,0 +1,416 @@
...
+namespace {
+  seastar::logger& logger() {
+    return crimson::get_logger(ceph_subsys_filestore);
+  }
+}
+
+
+namespace crimson::os::seastore::segment_manager::block {
+
+static write_ertr::future<> do_write(
```
[the remark this answers is not in the document]
Comment (athanatos) [reply src/crimson/os/seastore/segment_manager/block.cc:21]: Not if do_write and do_read are the only users.  If this becomes something useful in the future we can consider it.

File: src/crimson/os/seastore/segment_manager/block.cc (hunk 3 above)
Comment (tchaikov) [reply src/crimson/os/seastore/segment_manager/block.cc:296]: makes sense to me. by then we will be able to issue commands to query the segment states. could you note this down in `block.h` ? a very short note would suffice.

File: src/crimson/tools/store-nbd.cc [hunk 5]
```diff
@@ -0,0 +1,644 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:nil -*-
```
Comment (tchaikov) [inline src/crimson/tools/store-nbd.cc:1]: :+1:  personally, i think we should just use spaces in future. much easier when reviewing changes.

Comment (athanatos) [reply src/crimson/tools/store-nbd.cc:1]: Most of the code base does this.

File: src/crimson/os/seastore/journal.cc [hunk 6]
```diff
@@ -720,4 +734,110 @@ Journal::scan_segment_ret Journal::scan_segment(
...
+		cursor.offset);
+	      cursor.last_valid_header_found = true;
+	      return scan_valid_records_ertr::now();
+	    } else {
+	      logger().debug(
+		"Journal::scan_valid_records: valid record read at {}",
+		cursor.offset);
+	      cursor.last_committed = paddr_t{
+		cursor.offset.segment,
+		md->first.committed_to};
```
Comment (cyx1231st) [inline src/crimson/os/seastore/journal.cc:683]: I don't understand why need to persist `committed_to` in each `record_header_t`. According to implementation in https://github.com/ceph/ceph/pull/38290/commits/83414be599cd40af8848958cd502e720d4d04e84, it should be always pointing to the start offset of this record, correct?

Records are written in order, if we can make sure the next record's metadata is valid through nonce and crc checks, the current record should be committed already, right?

File: src/crimson/os/seastore/journal.cc [hunk 7]
```diff
@@ -600,37 +600,50 @@ Journal::replay_ret Journal::replay(delta_handler_t &&delta_handler)
...
+
+	  paddr_t extent_offset = base.add_offset(header.mdlength);
+	  for (const auto &i : *infos) {
+	    retref.emplace_back(extent_offset, i);
+	    extent_offset.offset += i.len;
+	  }
+	  return scan_extents_ertr::now();
+	}),
+      [&](auto &dhandler) {
+	return scan_valid_records(
```
Comment (cyx1231st) [inline src/crimson/os/seastore/journal.cc:639]: Is it a fatal error if any record becomes invalid during `scan_extents()`, infering data corruption from a dying SSD?

File: src/crimson/os/seastore/journal.cc [hunk 8]
```diff
@@ -403,28 +403,34 @@ Journal::find_replay_segments_fut Journal::find_replay_segments()
...
       try {
 	decode(header, bp);
       } catch (ceph::buffer::error &e) {
-	return read_record_metadata_ret(
-	  read_record_metadata_ertr::ready_future_marker{},
+	return read_validate_record_metadata_ret(
+	  read_validate_record_metadata_ertr::ready_future_marker{},
+	  std::nullopt);
+      }
+      if (header.segment_nonce != nonce) {
```
[the remark this answers is not in the document]
Comment (athanatos) [reply src/crimson/os/seastore/journal.cc:445]: Or from a previous write of the same segment on the same seastore.

File: src/crimson/os/seastore/journal.cc (hunk 6 above)
Comment (athanatos) [reply src/crimson/os/seastore/journal.cc:683]: No, the update to committed_to happens after the record commits.  Any record gets encoded with the address of the latest record to commit, which will be at least one record back and in the event that we've pipelined many of them possible more.  The purpose of committed_to is an optimization to avoid calculating data_crc except on the last few records of a segment.  If we only used this for replay, I wouldn't have bothered.  However, gc uses the same machinery, so it's in the fast path.

File: src/crimson/os/seastore/journal.cc (hunk 7 above)
Comment (athanatos) [reply src/crimson/os/seastore/journal.cc:639]: No, after a power-off event, we'd expect the tail records of any actively written segment to be invalid.

File: src/crimson/os/seastore/segment_manager/block.cc [hunk 9]
```diff
@@ -0,0 +1,407 @@
...
+Segment::close_ertr::future<> BlockSegment::close()
+{
+  manager.segment_close(id);
+  return close_ertr::now();
+}
+
+Segment::write_ertr::future<> BlockSegment::write(
+  segment_off_t offset, ceph::bufferlist bl)
+{
+  if (offset < write_pointer || offset % manager.superblock.block_size != 0)
```
[the remark this answers is not in the document]
Comment (athanatos) [reply src/crimson/os/seastore/segment_manager/block.cc:219]: Prior to performing a write, actually.  Will fix.

File: src/crimson/os/seastore/journal.cc (hunk 6 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/journal.cc:684]: Regardless of whether we've pipelined multiple records to submit or not. I'm wondering if the following pseudo code is identical to the current `scan_valid_records()` which can bypass most of data_crc checks and can work without `committed_to`:
```
INPUT: cursor_offset, nonce, budget, handler

current_md <- nil;
next_md <- nil;
budget_used <- 0;
if (cursor_offset == 0) {
    cursor_offset <- block_size
}
while (true) {
    current_md <- next_md
    next_md <- read_validate_record_metadata(cursor_offset, nonce)
    if (!next_md) {
        // the metadata of the next record is invalid, stop scanning
        break;
    } else if (current_md) {
        // the metadata of the next record is valid,
        // meaning the current record has completed submission,
        // thus no need to check data crc.
        handler(cursor_offset, current_md.header, current_md.mdbuffer)
        if (budget_used >= budget) {
            return budget_used
        }
    }
    record_length <- next_md.header.mdlength + next_md.header.dlength
    budget_used <- budget_used + record_length
    cursor_offset <- cursor_offset + record_length
}
if (current_md) {
    // the metadata of the next record is invalid,
    // meaning the current record is the final record and needs to check data crc.
    is_data_valid <- read_validate_data(cursor_offset, current_md.header)
    if (is_data_valid) {
        handler(cursor_offset, current_md.header, current_md.mdbuffer)
        record_length <- current_md.header.mdlength + current_md.header.dlength
        budget_used <- budget_used + record_length
    } else {
        // the current record is not valid, ignore
    }
} else {
    // the segment is empty.
}
return budget_used
```

Comment (athanatos) [reply src/crimson/os/seastore/journal.cc:684]: Your pseudocode works under the assumption that

   valid_metadata_crc_record(N+1) => valid_record(N)

This won't be true in general.  If we have N...N+5 in flight at the same time, the disk may persist the metadata for all of them without persisting the data for any of them.  The disk itself has no such ordering guarantee, and we don't necessarily wait for a flush on N before submitting N+1.

File: src/crimson/os/seastore/journal.cc (hunk 7 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/journal.cc:640]: I mean, do we need to worry about data corruption from a dying SSD in seastore?

Specifically, during `Journal::scan_extents()`, if unexpected bit flip is happened from the underlying device, causing crc mismatch, the current implementation just ignores the follow-up committed records in this segment and keeps on GC without error. Then later, a higher level application may crush due to the corrupted extents. And at that time, it could be very hard to distinguish whether it is a rear but fatal bug of the application, or it is a side-effect of data corruption.

File: src/crimson/os/seastore/journal.cc (hunk 6 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/journal.cc:684]: Thanks, it makes sense.

File: src/crimson/os/seastore/journal.cc (hunk 7 above)
Comment (athanatos) [reply src/crimson/os/seastore/journal.cc:640]: Yeah, that's potentially a problem.  At some point we'll probably want an independent registry of segment endpoints, but I'm leaving it for future work.

Comment (athanatos) [reply src/crimson/os/seastore/journal.cc:640]: More general checksumming in the normal read path will be via checksums embedded in the lba tree, btw.

**Review by cyx1231st — APPROVED** [review]
LGTM.

btw, I agree with Kefu that using spaces would make it easier to review complex logics from github, for example `Journal::scan_valid_records()` that contains long and nested `do_until()` loops.

**Comment by athanatos** [comment]
In that case, I think the step would be to submit a PR reformatting everything in crimson/ and fixing the emacs/vim modelines.

File: src/crimson/os/seastore/segment_manager/block.cc (hunk 2 above)
Comment (tchaikov) [reply src/crimson/os/seastore/segment_manager/block.cc:269]: i am sure. 


>        O_DSYNC
>               Write operations on the file will complete according to the
>               requirements of synchronized I/O data integrity completion.
> 
>               By the time write(2) (and similar) return, the output data has
>               been transferred to the underlying hardware, along with any
>               file metadata that would be required to retrieve that data
>               (i.e., as though each write(2) was followed by a call to
>               fdatasync(2)).  See NOTES below.
> 
> 

see also https://man7.org/linux/man-pages/man2/open.2.html

also, FWIW, seastar's implementation of `posix_file_impl::flush()` is a no-op if the file is opened with `open_flags::dsync`.

Comment (tchaikov) [reply src/crimson/os/seastore/segment_manager/block.cc:269]: also checked the linux kernel, it does respect the flag passed to `open()` when performing io using `file`.

see https://github.com/torvalds/linux/blob/cd796ed3345030aa1bb332fe5c793b3dddaf56e7/include/linux/fs.h#L3251 and https://github.com/torvalds/linux/blob/6ad4bf6ea1609fb539a62f10fca87ddbd53a0315/fs/block_dev.c#L199-L207

**Comment by tchaikov** [comment]
@athanatos i want to play around with the nbd based tool a little bit tomorrow before wrapping up the review.

File: src/crimson/os/seastore/journal.cc [hunk 10]
```diff
@@ -721,4 +735,108 @@ Journal::scan_segment_ret Journal::scan_segment(
...
+      return [=, &handler, &cursor, &budget_used] {
+	if (!cursor.last_valid_header_found) {
+	  return read_validate_record_metadata(cursor.offset, nonce
+	  ).safe_then([=, &cursor, &budget_used, &handler](auto md) {
+	    logger().debug(
+	      "Journal::scan_valid_records: read complete {}",
+	      cursor.offset);
+	    if (!md) {
+	      logger().debug(
+		"Journal::scan_valid_records: found invalid header at {}",
```
Comment (tchaikov) [inline src/crimson/os/seastore/journal.cc:674]: how about "header ends at {}" ? IMHO, "invalid" is always scaring..

File: src/crimson/os/seastore/journal.cc [hunk 11]
```diff
@@ -721,4 +735,108 @@ Journal::scan_segment_ret Journal::scan_segment(
...
+	  });
+	}
+      }().safe_then([=, &budget_used, &cursor] {
+	return scan_valid_records_ertr::make_ready_future<bool>(
+	  cursor.is_complete() || budget_used >= budget);
+      });
+    }).safe_then([retref=std::move(retref)]() mutable -> scan_valid_records_ret {
+      return scan_valid_records_ret(
+	scan_valid_records_ertr::ready_future_marker{},
+	std::move(*retref));
```
Comment (tchaikov) [inline src/crimson/os/seastore/journal.cc:750]: nit, `*retref` would suffice.

File: src/crimson/os/seastore/journal.h [hunk 12]
```diff
@@ -366,6 +366,55 @@ class Journal {
...
+  using scan_valid_records_ret = scan_valid_records_ertr::future<
+    size_t>;
+  using found_record_handler_t = std::function<
+    scan_valid_records_ertr::future<>(
+      paddr_t record_block_base,
+      // callee may assume header and bl will remain valid until
+      // returned future resolves
+      const record_header_t &header,
+      const bufferlist &bl)>;
+  scan_valid_records_ret scan_valid_records(
```
Comment (tchaikov) [inline src/crimson/os/seastore/journal.h:385]: might want to add a comment documenting what the return value is for. for instance
```c++
/// @returns the used budget
```

File: src/crimson/tools/store-nbd.cc [hunk 13]
```diff
@@ -0,0 +1,644 @@
...
+
+seastar::future<> NBDHandler::run()
+{
+  logger().debug("About to listen on {}", uds_path);
+  return seastar::do_with(
+    seastar::engine().listen(
+      seastar::socket_address{
+	seastar::unix_domain_addr{uds_path}}),
+    [=](auto &socket) {
+      return seastar::do_until(
```
[the remark this answers is not in the document]
Comment (athanatos) [reply src/crimson/tools/store-nbd.cc:437]: This permits the server to accept subsequent connections.

File: src/crimson/os/seastore/journal.cc (hunk 11 above)
Comment (athanatos) [reply src/crimson/os/seastore/journal.cc:750]: I'm not sure that's true.  *retref won't be an rvalue ref.

File: src/crimson/os/seastore/journal.cc [hunk 14]
```diff
@@ -315,7 +315,7 @@ Journal::find_replay_segments_fut Journal::find_replay_segments()
 	  } else {
 	    replay_from = paddr_t{from->first, (segment_off_t)block_size};
 	  }
-	  auto ret = std::vector<journal_seq_t>(segments.end() - from);
+	  auto ret = replay_segments_t(segments.end() - from);
```
[the remark this answers is not in the document]
Comment (athanatos) [reply src/crimson/os/seastore/journal.cc:401]: Different type.

**Comment by athanatos** [comment]
@tchaikov I've addressed/replied to comments.

**Comment by athanatos** [comment]
jenkins retest this please

**Comment by tchaikov** [comment]
> @tchaikov I've addressed/replied to comments.

ack. was working on the FTBFS yesterday. will finish the review today.

File: src/crimson/os/seastore/journal.cc [hunk 15]
```diff
@@ -601,37 +601,50 @@ Journal::replay_ret Journal::replay(delta_handler_t &&delta_handler)
...
+	  for (const auto &i : *infos) {
+	    retref.emplace_back(extent_offset, i);
+	    extent_offset.offset += i.len;
+	  }
+	  return scan_extents_ertr::now();
+	}),
+      [&](auto &dhandler) {
+	return scan_valid_records(
+	  cursor,
+	  header.segment_nonce,
```
Comment (tchaikov) [inline src/crimson/os/seastore/journal.cc:642]: @athanatos i am not sure about this part. `header` is a parameter of the `do_with()` block to scan the records with nonce in header. and it is captured by reference by the lambda passed to this `do_with()` block. the continuation could be scheduled to run after `header` is destructed if `scan_valid_records()` does not return a future which is not available immediately. so i'd suggest capture the nonce by a new value. like
```c++
      [&cursor, nonce=header.segment_nonce, this](auto &dhandler) {
        /// ...
```

File: src/crimson/os/seastore/journal.cc (hunk 11 above)
Comment (tchaikov) [reply src/crimson/os/seastore/journal.cc:750]: i mean,
```c++
return scan_valid_records_ertr::make_ready_future<size_t>(*retref);
```
is good enough. because `std::move(*retref)` makes user to believe that `*retref` is a heavy-weight object or it is a move-only object, so we should move it. but actually it is but the `budget_used` and it is a `size_t`.

File: src/crimson/os/seastore/journal.cc [hunk 16]
```diff
@@ -530,34 +530,57 @@ Journal::replay_segment(
...
 	}
+
+	return seastar::do_with(
+	  std::move(*deltas),
+	  [=](auto &deltas) {
+	    return crimson::do_for_each(
+	      deltas,
+	      [=](auto &delta) {
+		/* The journal may validly contain deltas for extents in
+		 * since released segments.  We can detect those cases by
```
Comment (tchaikov) [inline src/crimson/os/seastore/journal.cc:556]: nit, this comment is not part of this change, but would be great if you could take this opportunity to fix the syntax error by adding, for instance, "checking", so it reads

> We can detect those cases by checking whether ...

File: src/crimson/os/seastore/journal.cc [hunk 17]
```diff
@@ -404,28 +404,34 @@ Journal::find_replay_segments_fut Journal::find_replay_segments()
...
-  paddr_t start)
+Journal::read_validate_record_metadata_ret Journal::read_validate_record_metadata(
+  paddr_t start,
+  segment_nonce_t nonce)
 {
   if (start.offset + block_size > (int64_t)segment_manager.get_segment_size()) {
-    return read_record_metadata_ret(
-      read_record_metadata_ertr::ready_future_marker{},
+    return read_validate_record_metadata_ret(
+      read_validate_record_metadata_ertr::ready_future_marker{},
```
Comment (tchaikov) [inline src/crimson/os/seastore/journal.cc:427]: nit, FWIW, could just put
```c++
return {read_validate_record_metadata_ertr::ready_future_marker{},
        std::nullopt};
```
as C++17 is able to perform auto deduction from braced-init-list.

**Review by tchaikov — COMMENTED** [review]
just one thing, i cannot figure out why `header` can outlive that continuation.

File: src/crimson/tools/store-nbd.cc [hunk 18]
```diff
@@ -0,0 +1,642 @@
...
+	[this] { return false; },
```
Comment (tchaikov) [inline src/crimson/tools/store-nbd.cc:436]: could just put
```c++
return seastar::keep_doing(
```

File: src/crimson/os/seastore/journal.cc (hunk 15 above)
Comment (athanatos) [reply src/crimson/os/seastore/journal.cc:642]: Good catch!

File: src/crimson/os/seastore/journal.cc (hunk 11 above)
Comment (athanatos) [reply src/crimson/os/seastore/journal.cc:750]: Ok

File: src/crimson/os/seastore/journal.cc (hunk 17 above)
Comment (athanatos) [reply src/crimson/os/seastore/journal.cc:427]: There are actually a lot of lambdas in this file where the auto-deduction fails because there are multiple returns.  I'll keep this in mind in the future, but I'm not going to bother with it right now.

File: src/crimson/os/seastore/journal.cc (hunk 11 above)
Comment (athanatos) [reply src/crimson/os/seastore/journal.cc:750]: Actually, I'm going to leave it as is.  std::move here is technically correct, my intention is to permit a move construction as I do not mean to leave the previous location valid.  Doing it this way is type agnostic, if we change the return type here in the future to convey something more complex, std::move will remain correct while omitting the std::move will be incorrect.

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
@athanatos 
```
188/213 Test #208: unittest-seastore-journal .................***Failed   11.16 sec
```

might want to check the jenkins log for details.

**Comment by tchaikov** [comment]
test failure is tracked by https://tracker.ceph.com/issues/48108

**Comment by djgalloway** [comment]
All crimson builds started failing roughly around the time this PR got merged.  Could it be to blame for this:

```
Checking for unpackaged file(s): /usr/lib/rpm/check-files /home/jenkins-build/build/workspace/ceph-dev-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/16.0.0-8361-g63a5cd41/rpm/el8/BUILDROOT/ceph-16.0.0-8361.g63a5cd41.el8.x86_64
error: Installed (but unpackaged) file(s) found:
   /usr/bin/crimson-store-nbd
   /usr/lib/debug/usr/bin/crimson-store-nbd-16.0.0-8361.g63a5cd41.el8.x86_64.debug


RPM build errors:
    Duplicate build-ids /home/jenkins-build/build/workspace/ceph-dev-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/16.0.0-8361-g63a5cd41/rpm/el8/BUILDROOT/ceph-16.0.0-8361.g63a5cd41.el8.x86_64/usr/bin/ceph-osd and /home/jenkins-build/build/workspace/ceph-dev-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/16.0.0-8361-g63a5cd41/rpm/el8/BUILDROOT/ceph-16.0.0-8361.g63a5cd41.el8.x86_64/usr/bin/crimson-osd
    Duplicate build-ids /home/jenkins-build/build/workspace/ceph-dev-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/16.0.0-8361-g63a5cd41/rpm/el8/BUILDROOT/ceph-16.0.0-8361.g63a5cd41.el8.x86_64/usr/lib/debug/usr/bin/ceph-osd-16.0.0-8361.g63a5cd41.el8.x86_64.debug and /home/jenkins-build/build/workspace/ceph-dev-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/16.0.0-8361-g63a5cd41/rpm/el8/BUILDROOT/ceph-16.0.0-8361.g63a5cd41.el8.x86_64/usr/lib/debug/usr/bin/crimson-osd-16.0.0-8361.g63a5cd41.el8.x86_64.debug
    Installed (but unpackaged) file(s) found:
   /usr/bin/crimson-store-nbd
   /usr/lib/debug/usr/bin/crimson-store-nbd-16.0.0-8361.g63a5cd41.el8.x86_64.debug
+ rm -fr /tmp/install-deps.2911679
Build step 'Execute shell' marked build as failure
```

**Comment by athanatos** [comment]
@djgalloway Yeah, https://github.com/ceph/ceph/pull/38608 is an attempt to fix it, but it seems to be failing make check for unrelated reasons.
